rs/boundary_node/salt_sharing/canister/canister.rs

```diff
@@ -0,0 +1,161 @@
+use std::cell::RefCell;
+use std::time::Duration;
+
+use crate::storage::{LAST_SALT_NS, SALT, SALT_SIZE};
+use crate::time::delay_till_next_month;
+use candid::Principal;
+use ic_cdk::{api::time, spawn};
+use ic_cdk_macros::{init, post_upgrade, query};
+use ic_cdk_timers::set_timer;
+use rand_chacha::{
+    rand_core::{RngCore, SeedableRng},
+    ChaCha20Rng,
+};
+use salt_api::{GetSaltError, GetSaltResponse, InitArg, SaltGenerationStrategy, SaltResponse};
+
+// A small delay after beginning of the month when salt is generated
+const START_OF_MONTH_DELAY: Duration = Duration::from_secs(5);
+// Maximum number of attempts to initialize random seed
+const MAX_SEED_INIT_ATTEMPTS: i32 = 15;
+// Interval between retries to initialize random seed
+const SEED_INIT_RETRY_INTERVAL: Duration = Duration::from_secs(10);
+
+thread_local! {
+    // Random seed used for salt generation
+    static RNG: RefCell<Option<ChaCha20Rng>> = const { RefCell::new(None) };
+}
+
+// Run when the canister is first installed
+#[init]
+fn init(init_arg: InitArg) {
+    // Generate salt on the very first init or based on the provided argument.
+    if !is_salt_init() || init_arg.regenerate_now {
+        seed_rng_and_regenerate_salt(MAX_SEED_INIT_ATTEMPTS, Duration::ZERO);
+    }
+    // Start salt generation schedule based on the argument.
+    if let Some(strategy) = init_arg.salt_generation_strategy {
+        schedule_salt_generation(strategy);
+    }
+    // Start schedule of seeding randomness, if seeding_interval is None, schedule is executed only once.
+    let seeding_interval = init_arg.seeding_interval_secs.map(Duration::from_secs);
+    schedule_rng_seeding(Duration::ZERO, seeding_interval);
+}
+
+// Run every time a canister is upgraded
+#[post_upgrade]
+fn post_upgrade(init_arg: InitArg) {
+    // Run the same initialization logic
+    init(init_arg);
+}
+
+#[query]
+fn get_salt() -> GetSaltResponse {
+    get_salt_response()
```

Comment:
should there be some access control for retrieving the salt?

Reply:
Yes, surely, i will add it soon, in next PR.

Reply:
Is this tracked somewhere?

Reply:
yes, i have a planning doc for this feature and will make tickets out of it